src/datasets/iterable_dataset.py

```diff
@@ -281,6 +281,32 @@ def n_shards(self) -> int:
         return self.ex_iterable.n_shards
 
 
+class TypedExamplesIterable(_BaseExamplesIterable):
+    def __init__(self, ex_iterable: _BaseExamplesIterable, features: Features):
+        self.ex_iterable = ex_iterable
+        self.features = features
+
+    def __iter__(self):
+        iterator = iter(self.ex_iterable)
+        for key, example in iterator:
```

Comment:
(nit) It's maybe simpler this way
```suggestion
        for key, example in self.ex_iterable:
```